python-runnables/import-sheets-into-project/runnable.py

```diff
@@ -0,0 +1,165 @@
+        """
+        logger.info("GoogleSheets macro v{} starting with {} on project {}".format(DSSConstants.PLUGIN_VERSION, logger.filter_secrets(config), project_key))
+        self.project_key = project_key
+        self.config = config
+        self.is_dry_run = self.config.get("is_dry_run", True)
+        self.plugin_config = plugin_config
+        self.doc_id = self.config.get("doc_id")
+        self.tabs_ids = self.config.get("tabs_ids", [])
+        self.result_format = self.config.get("result_format")
+        self.write_format = self.config.get("write_format")
+        credentials, credentials_type = extract_credentials(config)
+        self.session = GoogleSheetsSession(credentials, credentials_type)
+        dss_client = dataiku.api_client()
+        self.project = dss_client.get_project(project_key)
+        self.project_datasets = list_project_datasets_names(self.project)
+        self.creation_mode = self.config.get("creation_mode", "create-new")
+        self.worksheets = self.session.get_spreadsheets(self.doc_id)
+        if not self.tabs_ids:
+            for worksheet in self.worksheets:
+                self.tabs_ids.append(worksheet.title)
+
+    def get_progress_target(self):
+        """
+        If the runnable will return some progress info, have this function return a tuple of 
+        (target, unit) where unit is one of: SIZE, FILES, RECORDS, NONE
+        """
+        return (len(self.tabs_ids), 'FILES')
+
+    def run(self, progress_callback):
+        """
+        Do stuff here. Can return a string or raise an exception.
+        The progress_callback is a function expecting 1 value: current progress
+        """
+        spreadsheet_title = self.session.get_spreadsheet_title(self.doc_id)
+        if not spreadsheet_title:
+            spreadsheet_title = "Nameless spreadsheet"
+        project_flow = self.project.get_flow()
+
+        result_table = ResultTable()
+        result_table.add_column("actions", self._get_text("actions"), "STRING")
+
+        target_zone = get_zone_from_name(project_flow, spreadsheet_title)
+        if not target_zone:
+            result_table.add_record([self._get_text("creating").format(spreadsheet_title=spreadsheet_title)])
+            if not self.is_dry_run:
+                target_zone = project_flow.create_zone(spreadsheet_title)
+
+        if self.creation_mode == "create-new":
+            worksheets_titles = self.project_datasets.copy()
+        else:
+            worksheets_titles = []
+        index = 0
+        for worksheet in self.worksheets:
+            index += 1
+            progress_callback(index)
+            dataset = None
+            worksheet_title = worksheet.title
+            worksheets_titles.append("{}_{}".format(spreadsheet_title, worksheet_title))
+            worksheets_titles = get_unique_slugs(worksheets_titles)
+            unique_worksheet_title = worksheets_titles[-1:][0]
+
+            if worksheet_title in self.tabs_ids:
```

Comment:
If I'm not mistaken, this `if` should be moved to line 72, just after the `for`, otherwise we would (among other things) increment `index` for each and every worksheet, even the ones not selected by the user for import. Currently the user sees wrong progress report:
![image](https://user-images.githubusercontent.com/17353043/215558003-3024d120-50b9-4bb9-863c-a94d89d3ed88.png)

Reply:
Done in [607299a](https://github.com/dataiku/dss-plugin-googlesheets/pull/7/commits/607299a6a30aba743cacbbee9c55ee6c2668c44e)